misc/python/mzbuild.py

```diff
@@ -0,0 +1,359 @@
+# Copyright Materialize, Inc. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+from functools import lru_cache
+from collections import OrderedDict
+from pathlib import Path
+from tempfile import TemporaryFile
+from typing import cast, Any, Dict, List, Set, Optional, Union, Iterable, Iterator, IO
+import base64
+import contextlib
+import enum
+import hashlib
+import json
+import os
+import re
+import shutil
+import stat
+import subprocess
+import sys
+import yaml
+
+
+class AcquiredFrom(enum.Enum):
+    REGISTRY = "registry"
+    LOCAL_BUILD = "local-build"
+
+
+def runv(args: List[Union[Path, str]],
+         cwd: Optional[Path] = None,
+         stdin: Optional[IO[bytes]] = None) -> None:
+    args = [str(arg) for arg in args]  # type: List[str]
+    cwd = str(cwd) if cwd is not None else None  # type: Optional[str]
+    print("$", " ".join(args))
+    subprocess.check_call(args, cwd=cwd, stdin=stdin)
+
+
+def capture(args: List[Union[Path, str]],
+            cwd: Optional[Path] = None,
+            universal_newlines: bool = False) -> Any:
+    args = [str(arg) for arg in args]  # type: List[str]
+    cwd = str(cwd) if cwd is not None else None  # type: Optional[str]
+    return subprocess.check_output(args,
+                                   cwd=cwd,
+                                   universal_newlines=universal_newlines)
+
+
+def xcargo(root: Path) -> Path:
+    if sys.platform == "linux":
+        return "cargo"
+    else:
+        return root / "bin" / "xcompile"
+
+
+def xcargo_target_dir(root: Path) -> Path:
+    if sys.platform == "linux":
+        return root / "target"
+    else:
+        return root / "target" / "x86_64-unknown-linux-gnu"
+
+
+def xstrip(root: Path) -> str:
+    if sys.platform == "linux":
+        return "strip"
+    else:
+        return "x86_64-unknown-linux-gnu-strip"
+
+
+def docker_images() -> Set[str]:
+    return set(
+        capture(["docker", "images", "--format", "{{.Repository}}:{{.Tag}}"],
+                universal_newlines=True).strip().split("\n"))
+
+
+def git_ls_files(root: Path, *specs: str) -> List[bytes]:
+    files = capture([
+        "git", "ls-files", "--cached", "--others", "--exclude-standard", "-z",
+        *specs
+    ],
+                    cwd=root).split(b"\0")
+    return [f for f in files if f.strip() != b""]
+
+
+def chmod_x(path: Path) -> None:
+    # https://stackoverflow.com/a/30463972/1122351
+    mode = os.stat(str(path)).st_mode
+    mode |= (mode & 0o444) >> 2  # copy R bits to X
+    os.chmod(str(path), mode)
+
+
+class PreImage:
+    def run(self, root: Path, path: Path) -> None:
+        runv(["git", "clean", "-ffdx", path])
```

Comment:
I don't know if it should be possible to delete all untracked files by running a build command? What context makes this safe?

Reply:
agree, I don't like build tools that run `git clean -dfx` or in any other way change things that they didn't create.

Reply:
Oops, I guess I wanted `git clean -ffX` to only remove the .gitignored files. The reason this is safe is because the clean is scoped to an "mzbuild context", where the rules are: if you've gitignored a file in this directory, it is something that mzbuild is responsible for creating, and something that is safe to delete at the beginning of the build process. I've added a note about this to do the docs.